libraries/AC_WPNav/AC_WPNav.cpp

```diff
@@ -374,158 +448,134 @@ void AC_WPNav::get_wp_stopping_point(Vector3f& stopping_point) const
+    update_pos_vel_accel(_pos_terrain_offset, _vel_terrain_offset, _accel_terrain_offset, _track_scalar_dt * dt);
+
     // convert final_target.z to altitude above the ekf origin
-    final_target.z += terr_offset;
-    _pos_control.set_pos_target(final_target);
+    target_pos.z += _pos_terrain_offset;
+    target_vel.z += _vel_terrain_offset;
+    target_accel.z += _accel_terrain_offset;
+
+    // pass new target to the position controller
+    _pos_control.set_pos_vel_accel_target(target_pos, target_vel, target_accel);
 
     // check if we've reached the waypoint
-    if( !_flags.reached_destination ) {
-        if( _track_desired >= _track_length ) {
+    if (!_flags.reached_destination) {
+        if (s_finished) {
             // "fast" waypoints are complete once the intermediate point reaches the destination
             if (_flags.fast_waypoint) {
                 _flags.reached_destination = true;
-            }else{
+            } else {
                 // regular waypoints also require the copter to be within the waypoint radius
-                Vector3f dist_to_dest = (curr_pos - Vector3f(0,0,terr_offset)) - _destination;
-                if( dist_to_dest.length() <= _wp_radius_cm ) {
+                const Vector3f dist_to_dest = curr_pos - _destination;
+                if (dist_to_dest.length() <= _wp_radius_cm) {
                     _flags.reached_destination = true;
                 }
             }
         }
     }
 
-    // update the target yaw if origin and destination are at least 2m apart horizontally
-    if (_track_length_xy >= WPNAV_YAW_DIST_MIN) {
-        if (_pos_control.get_leash_xy() < WPNAV_YAW_DIST_MIN) {
-            // if the leash is short (i.e. moving slowly) and destination is at least 2m horizontally, point along the segment from origin to destination
-            set_yaw_cd(get_bearing_cd(_origin, _destination));
-        } else {
-            Vector3f horiz_leash_xy = final_target - curr_pos;
-            horiz_leash_xy.z = 0;
-            if (horiz_leash_xy.length() > MIN(WPNAV_YAW_DIST_MIN, _pos_control.get_leash_xy()*WPNAV_YAW_LEASH_PCT_MIN)) {
-                set_yaw_cd(RadiansToCentiDegrees(atan2f(horiz_leash_xy.y,horiz_leash_xy.x)));
-            }
+    // Calculate the turn rate
+    float turn_rate = 0.0f;
+    const float target_vel_xy_len = Vector2f(target_vel.x, target_vel.y).length();
+    if (is_positive(target_vel_xy_len)) {
+        const float accel_forward = (target_accel.x * target_vel.x + target_accel.y * target_vel.y + target_accel.z * target_vel.z)/target_vel.length();
+        const Vector3f accel_turn = target_accel - target_vel * accel_forward / target_vel.length();
+        const float accel_turn_xy_len = Vector2f(accel_turn.x, accel_turn.y).length();
+        turn_rate = accel_turn_xy_len / target_vel_xy_len;
+        if ((accel_turn.y * target_vel.x - accel_turn.x * target_vel.y) < 0.0) {
+            turn_rate *= -1.0f;
```

Comment:
```suggestion
            turn_rate = -turn_rate;
```

Reply:
replace the multiply with an assignment.